tests/Add-DbaAgDatabase.Tests.ps1

```diff
@@ -38,49 +43,122 @@ Describe "Add-DbaAgDatabase" -Tag "UnitTests" {
+        $PSDefaultParameterValues['*-Dba*:EnableException'] = $true
+
+        # Collect all the created files to be able to remove them in the AfterAll block.
+        $filesToRemove = @( )
+
+        # Explain what needs to be set up for the test:
+        # To add a database to an availablity group, we need an availability group and a database that has been backed up.
+        # For negative tests, we need a database without a backup and a non existing database.
+
+        # Set variables. They are available in all the It blocks.
+        $agName = "addagdb_group"
+        $existingDbWithBackup = "dbWithBackup"
+        $existingDbWithoutBackup = "dbWithoutBackup"
+        $nonexistingDb = "dbdoesnotexist"
+
+        # Create the objects.
+        $splat = @{
+            Primary      = $TestConfig.instance3
+            Name         = $agName
+            ClusterType  = "None"
             FailoverMode = "Manual"
-            Database = $dbname
-            Confirm = $false
-            Certificate = "dbatoolsci_AGCert"
+            Certificate  = "dbatoolsci_AGCert"
         }
-        $ag = New-DbaAvailabilityGroup @splatNewAg
+        $null = New-DbaAvailabilityGroup @splat
+
+        $null = New-DbaDatabase -SqlInstance $TestConfig.instance3 -Name $existingDbWithBackup
+        $backup = Backup-DbaDatabase -SqlInstance $TestConfig.instance3 -Database $existingDbWithBackup -Path $TestConfig.Temp
+        $filesToRemove += $backup.Path
+
+        $null = New-DbaDatabase -SqlInstance $TestConfig.instance3 -Name $existingDbWithoutBackup
+
+        # We want to run all commands outside of the BeforeAll block without EnableException to be able to test for specific warnings.
+        $PSDefaultParameterValues.Remove('*-Dba*:EnableException')
     }
 
     AfterAll {
-        $null = Remove-DbaAvailabilityGroup -SqlInstance $server -AvailabilityGroup $agname -Confirm:$false
-        $null = Remove-DbaDatabase -SqlInstance $server -Database $dbname, $newdbname -Confirm:$false
+        # We want to run all commands in the AfterAll block with EnableException to ensure that the test fails if the cleanup fails.
+        $PSDefaultParameterValues['*-Dba*:EnableException'] = $true
+
+        # Cleanup all created object.
+        $null = Remove-DbaAvailabilityGroup -SqlInstance $TestConfig.instance3 -AvailabilityGroup $agName
+        $null = Get-DbaEndpoint -SqlInstance $TestConfig.instance3 -Type DatabaseMirroring | Remove-DbaEndpoint
+        $null = Remove-DbaDatabase -SqlInstance $TestConfig.instance3 -Database $existingDbWithBackup, $existingDbWithoutBackup
+
+        # Remove all created files.
+        Remove-Item -Path $filesToRemove
+
+        # As this is the last block we do not need to reset the $PSDefaultParameterValues.
     }
 
     Context "When adding AG database" {
+        # We use the BeforeAll to run the test itself.
+        # Results are saved in $results.
+        BeforeAll {
+            $splat = @{
```

Comment:
I just named all splatting hashtables $splat as the variable is only used once just after the definition. This way we don't need to come up with all the names for the splats.